internal/command/init.go

```diff
@@ -262,24 +262,22 @@ func (c *InitCommand) Run(args []string) int {
 		return 1
 	}
 
-	// If we pass the core version check, we want to show any errors from initializing the backend next,
-	// which will include syntax errors from loading the configuration. However, there's a special case
-	// where we are unable to load the backend from configuration or state _and_ the configuration has
-	// errors. In that case, we want to show a slightly friendlier error message for newcomers.
-	showBackendDiags := back != nil || rootModEarly.Backend != nil || rootModEarly.CloudConfig != nil
-	if showBackendDiags {
-		diags = diags.Append(backDiags)
-		if backDiags.HasErrors() {
-			c.showDiagnostics(diags)
-			return 1
-		}
-	} else {
-		diags = diags.Append(earlyConfDiags)
-		if earlyConfDiags.HasErrors() {
-			c.Ui.Error(strings.TrimSpace(errInitConfigError))
-			c.showDiagnostics(diags)
-			return 1
-		}
+	// We've passed the core version check, now we can show errors from the
+	// configuration and backend initialisation.
+
+	// Now, we can check the diagnostics from the early configuration.
+	diags = diags.Append(earlyConfDiags)
+	if earlyConfDiags.HasErrors() {
```

Comment:
Hmm, `earlyConfDiags` doesn't include errors from backend initialization, so if I have a config that looks like this:

```
terraform {
  backend "local" {
    bad_attribute = "is bad"
  }
}

bad_block {

}
```

I won't see anything about the backend problems until after I fix any other config errors. 

That doesn't seem like a terrible outcome, but I wonder if it might be just as well to return all these diags together and adjust `errInitConfigError` so that it makes sense for both config errors and backend errors - the exact wording of that error message used to be important because `terraform-exec` relied on it, but that matching was since [removed](https://github.com/hashicorp/terraform-exec/pull/352).

Reply:
Yep, sounds good! I've made it print both sets if either one fails and tweaked the wording a bit!

Reply:
Ah, actually, an important bit of context came back to me and I fear I gave you a bad suggestion 😅 `backDiags` includes config validation errors as well, so appending/returning both together means we'll show errors twice.

We could consider:

1. making `loadBackendConfig` _not_ also [validate the entire root module](https://github.com/hashicorp/terraform/blob/c9a7b7f082551da9ea2b18a326009573487d60d0/internal/command/meta_config.go#L130-L144) so `backDiags` actually contains what it says on the tin
2. going back to what you had originally in order to fix the crash and straighten out this initialization logic later

2 seems reasonable to me, unless you have other suggestions - sorry for the false flag!

Reply:
I'm more than happy to turn it back, but I don't think they will actually contain duplicate diagnostics. The HCL library itself caches the files it loads: https://github.com/hashicorp/hcl/blob/main/hclparse/parser.go#L56

So, the first time round it loads the partial file with the diagnostics, but on following reads it doesn't reproduce the diagnostics. It just returns the partial file directly with no diagnostics. The tests also don't show duplicate errors.

Either way, I'm more than happy to go back to the previous approach but I *think* it should be fine.